squint.c

```diff
@@ -0,0 +1,1611 @@
+int *rethread_branch(int *branchInst)
+{
+   int* retVal;
+   if  ((*branchInst & 0xff000000) == 0xea000000) { // uncond branch
+      int tmp = (*branchInst & 0x00ffffff) |
+                ((*branchInst & 0x00800000) ? 0xff000000 : 0);
+
+      int *dstInst = branchInst + 2 + tmp;
+      retVal = rethread_branch(dstInst);
+      *branchInst += (retVal - dstInst);
+   }
+   else {
+      retVal = branchInst;
+   }
+   return retVal;
+}
+
+/* rethread unconditional branch chains */
+void simplify_branch2(int *funcBegin, int *funcEnd) {
+   int *scan;
+   for (scan = funcBegin; scan <= funcEnd; ++scan) {
+      scan = skip_nop(scan, 1);
+      if ((*scan & 0xff000000) == 0xea000000) { // uncond branch
+         rethread_branch(scan);
+      }
+   }
+}
+
+/* amacc has condtional branches that jump to a compare instruction */
+/* follow chains of these instructions to go directly to final address */
+void simplify_branch3(int *funcBegin, int *funcEnd) {
+   int *scan;
+   int match, tmp;
+   for (scan = funcBegin; scan < funcEnd; ++scan) {
+      scan = skip_nop(scan, 1);
+      if (*scan == 0xe3500000) { /* cmp r0, #0 */
+         int *branchInst = active_inst(scan, 1);
+         int *target = branchInst;
+         if ((*target & 0xff000000) == 0x0a000000 || /* beq */
+             (*target & 0xff000000) == 0x1a000000) { /* bne */
+            match = (*target & 0xf0000000);
+            while(1) {
+               /* calc sign extended branch addr */
+               tmp = (*target & 0x00ffffff) |
+                     ((*target & 0x00800000) ? 0xff000000 : 0);
+               target += 2 + tmp;
+               if (*target == 0xe3500000) { /* cmp r0, #0 */
+                  target = active_inst(target, 1);
+               }
+               if ((*target & 0xff000000) == 0x0a000000 || /* beq */
+                   (*target & 0xff000000) == 0x1a000000) { /* bne */
+                  if ((*target & 0xf0000000) != match) {
+                     ++target;
+                     if ((*target & 0xff000000) != 0xea000000) { /* b */
+                        break;
+                     }
+                  }
+               }
+               else if ((*target & 0xff000000) != 0xea000000) { /* b */
+                  break;
+               }
+            }
+            *branchInst = (*branchInst & 0xff000000) |
+                          ((target - branchInst - 2) & 0xffffff);
+         }
+      }
+   }
+}
+
+/* Now that we have simplified all final branch targets, we */
+/* will simplify and compress out extraneous comparison opcodes. */
+void simplify_branch4(int *funcBegin, int *funcEnd)
+{
+   int *scan;
+   for (scan = funcBegin; scan < funcEnd; ++scan) {
+      scan = skip_nop(scan, 1);
+
+      if (*scan == 0xe3500000) { /* cmp r0, #0 */
+         int *scanm1, *scanm2, *scanm3, *scanp1;
+
+         /* simplify cmp blocks */
+         scanm1 = active_inst(scan,   -1);
+         scanm2 = active_inst(scanm1, -1);
```

Comment:
I don't follow the logic of this function.  Why can we apply the optimization without checking the instruction that `scanm2` and `scanp1` point to?

Reply:
Line 1596 along with lines 1715-1722  generate a four instruction sequence for comparisons, where *scanm3 is "cmp rX, r0" and *scan is "cmp r0, #0".  AMaCC literally cannot generate these two specific compares in any situation other than a four instruction block, so no need for checks.  As I discuss in docs/Squint.md, Squint is not a typical peephole optimizer, because it leverages idioms produced by the compiler, and is thus much more powerful.  In fact, Squint cuts the number of executable instructions roughly in half, better than any other peephole optimizer I am aware of.  I think this comment suggests that Squint should behave like a small window peephole optimizer which is much than what Squint can do.  I believe that decoupling Squint from idioms produced by the compiler goes against the philosopy of squint.  I agree that if the compiler changes, squint must change with it, but I also think it is worth the end result of cutting executable instructions in half!!

This response applies to your comments about the apply_peepholesN to some degree, but not as strongly.  The current peephole functions can not run into instruction widow problems, beause the conditionals being checked will cause an early termination of a peephole optimization before a full window worth of instructions need to be checked.  Checking for bounds more carefully can double the amount of processong needed to apply the peephole  operations, so you need to trade off that extra cost against the fact that a window bounds violation can never happen, given the way that the code is currently written.  Right now, Squint takes a very small amount of time, but as other optimizations are added, that time can increase significantly, so paying the price of doubling that longer processing time that may occur in the future is not something I feel comfortable building into Squint from the outset, when it is not needed.

Reply:
@jserv You may need to comment on this as a tie-breaker.

Reply:
@loganchien "The active_inst on Line 730-732 risks out-of-bound array access".  Look more carefully. As I discussed in my previous comment, the comparison on lines 725-727 make this a non-issue, because when AMaCC generates a "pop {R1}" instruction, it is guaranteed to generate at least one instruction after that, and when you add that to the two instructions that are generated by LEV, no violation can occur.

Reply:
Final comment: sieve.c is the only "benchmark" test in the repo, and it runs ~3.5x faster due to Squint philosophy!  Name another peephole optimizer that is that good!

Reply:
Let me approach this from another angle.  Let's say that the idiom AMaCC uses to generate comparison instructions changes so that the  four instruction comparison block goes away. Or assume that that the "push r0; ...; pop {r1}; <action>" idiom goes away.  In either case, the idiom AMaCC generates to replace those idioms is unlikely to be similar enough to match one of the complex peephole optimization patterns.  What is most likely to happen is that because the criterion is not met, the optimization will not be applied.  I would bet that essentially always the outcome would either be "do no harm, and don't apply the optimization" or "catastrophic error".  The transformation will likely go so wrong, that the test suite will immediately catch it, causing all tests with the optimizer turned on to fail.  In my mind, the fact that either one extreme or the other is almost certain to happen is the best way to mitigate risk, given that a test suite run is required before a merge.

Reply:
The compiler developers prefer the conservative approach because the compiler must accept all well-formed input program and is expected to generate correct program.  That's why the compiler developers try hard to proof the correctness of all compiler transformations and minimize the contextual assumption (unless they can proof it properly).  If there is a pattern that is very critical and shouldn't be missed by the optimizer, people add an end-to-end test case to check whether the pattern is properly optimized.

In the conservative approach, the compiler developer clearly knows which pattern will be transformed and knows how to write a test to ensure the transform is applied.  In the aggressive approach, the compiler developer doesn't know whether an input program may have a matching pattern that should not be transformed.  Even with a large set of test cases, we can only claim that bug doesn't appear in the test cases.  We cannot claim that the compiler works for all inputs.  Not even mentioning that developing a good C programming language test suite is not a trivial task and we only have a small set of test cases in the amacc repository.

> Squint is not a typical peephole optimizer, because it leverages idioms produced by the compiler, and is thus much more powerful.

I don't see why properly checking the pre-condition before applying the optimization will stop Squint from reducing the number of instruction in the output executable.

I don't see the benefits of coupling the optimizer with the code generation implementation details either.  If more transformations are developed, the program under transformation will gradually be different from what the code generator initially generated.

>  Checking for window bounds more carefully, as you are suggesting, can double the amount of processing needed to apply the peephole operations

It depends on how you implement the check.  You don't have to perform the check in the loop.  For example, for `apply_peepholes2`, you can write your program as:

```
for (int i = 0; i < 2; ++i) {
  scan = active_inst(funcBegin, 1);
  if (scan > funcEnd) return ...;
}

for (int i = 0; i < 6; ++i) {
  scanEnd = active_inst(funcEnd, -1);
  if (scanEnd <= scan) return;
}

for (; scan < scanEnd; ++scan) {
  scan = skip_nop(scan, S_FWD);
  // ... original logic ...
}
```

Reply:
"The compiler developers prefer the conservative approach... That's why the compiler developers try hard to proof the correctness of all compiler transformations"

Really?  AMaCC has literally dozens of bugs.  Some that I have fixed in previous updates, some that I have submitted issues for, and many that I have not logged at all but I am aware of.  Why is it that my update has to achieve absolute perfection, when clearly there was almost no scrutiny by reviewers at all for previous updates to AMaCC?  Is this fair? Serious question.
Everything I have committed in my past AMaCC pull requests have been not only rock solid but extremely compact (i.e. exceptionally good code) and Squint is solid (other than one bug I admitted to early in this thread in relocate_nop and one other that I am aware of and plan to fix within a week or two that no one will even encounter).  We can talk about theoreretical bugs in this code all day, but in terms of actual bugs, there are two, and they are known and unimportant to the current function of Squint. This code works, is very good, and I'm not changing it.  You can change it after the update if you want.  PS  Squint is experimental and not even part of the compiler, making the fact that you are holding me to ridiculous high standard that no other developer had to be even remotely close to even more unfair.  If you want to make changes after the pull completes, you are welcome to do so.

Reply:
PS  I believe what I am doing is the right approach, and I also believe it is conservative, for the reasons I tried to explain at length.  It a matter of coding philosopy, not bugs, because there are no bugs (other than the two that are known and not relevant in terms of the code development strategy).

Reply:
"I don't see the benefits of coupling the optimizer with the code generation implementation details"

Have you actually looked at the quality of optimizations using scripts/disasm on amacc.c before and after the optimization?  If you haven't you need to do so. The executable size is cut in half (ignoring NOPS), and the optimizations are excellent.  The end result is achieved in so few lines of peephole optimization *only* because I focused on idioms generated by the compiler.   PS sieve.c runs about 3.5x faster, so the performance optimization is not bad either!